File: TarsEngine.FSharp.Core/TreeOfThought/ITreeOfThoughtService.fs
namespace TarsEngine.FSharp.Core.TreeOfThought

open System.Threading.Tasks

/// <summary>
/// Interface for Tree-of-Thought service.
/// </summary>
type ITreeOfThoughtService =
    /// <summary>
    /// Creates a thought tree for the specified problem.
    /// </summary>
    /// <param name="problem">The problem to solve.</param>
    /// <param name="options">The options for tree creation.</param>
    /// <returns>The thought tree.</returns>
    abstract member CreateThoughtTreeAsync : problem:string * options:TreeCreationOptions -> Task<IThoughtNode>
    
    /// <summary>
    /// Evaluates a thought node.
    /// </summary>
    /// <param name="node">The node to evaluate.</param>
    /// <param name="metrics">The evaluation metrics.</param>
    /// <returns>The evaluated node.</returns>
    abstract member EvaluateNodeAsync : node:IThoughtNode * metrics:EvaluationMetrics -> Task<IThoughtNode>
    
    /// <summary>
    /// Adds a child to a node.
    /// </summary>
    /// <param name="parent">The parent node.</param>
    /// <param name="childThought">The child thought.</param>
    /// <returns>The updated parent node.</returns>
    abstract member AddChildAsync : parent:IThoughtNode * childThought:string -> Task<IThoughtNode>
    
    /// <summary>
    /// Selects the best node from a thought tree.
    /// </summary>
    /// <param name="root">The root node.</param>
    /// <returns>The best node.</returns>
    abstract member SelectBestNodeAsync : root:IThoughtNode -> Task<IThoughtNode>
    
    /// <summary>
    /// Prunes nodes that don't meet a threshold.
    /// </summary>
    /// <param name="root">The root node.</param>
    /// <param name="threshold">The threshold.</param>
    /// <returns>The pruned tree.</returns>
    abstract member PruneByThresholdAsync : root:IThoughtNode * threshold:float -> Task<IThoughtNode>
    
    /// <summary>
    /// Generates a report for a thought tree.
    /// </summary>
    /// <param name="root">The root node.</param>
    /// <param name="title">The report title.</param>
    /// <returns>The report.</returns>
    abstract member GenerateReportAsync : root:IThoughtNode * title:string -> Task<string>
    
    /// <summary>
    /// Saves a report for a thought tree.
    /// </summary>
    /// <param name="root">The root node.</param>
    /// <param name="title">The report title.</param>
    /// <param name="filePath">The file path.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    abstract member SaveReportAsync : root:IThoughtNode * title:string * filePath:string -> Task

File: archive/v1/parked_csharp/TarsEngine.CSharp.Adapters/TreeOfThought/ITreeOfThoughtService.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace TarsEngine.CSharp.Adapters.TreeOfThought
{
    /// <summary>
    /// Interface for the Tree of Thought service.
    /// </summary>
    public interface ITreeOfThoughtService
    {
        /// <summary>
        /// Creates a root node with the specified thought.
        /// </summary>
        /// <param name="thought">The thought content.</param>
        /// <returns>The root node.</returns>
        Task<ThoughtNodeAdapter> CreateRootNodeAsync(string thought);

        /// <summary>
        /// Adds a child node to the specified parent node.
        /// </summary>
        /// <param name="parent">The parent node.</param>
        /// <param name="thought">The thought content.</param>
        /// <returns>The updated parent node.</returns>
        Task<ThoughtNodeAdapter> AddChildNodeAsync(ThoughtNodeAdapter parent, string thought);

        /// <summary>
        /// Adds multiple child nodes to the specified parent node.
        /// </summary>
        /// <param name="parent">The parent node.</param>
        /// <param name="thoughts">The thought contents.</param>
        /// <returns>The updated parent node.</returns>
        Task<ThoughtNodeAdapter> AddChildNodesAsync(ThoughtNodeAdapter parent, IEnumerable<string> thoughts);

        /// <summary>
        /// Evaluates the specified node with the specified metrics.
        /// </summary>
        /// <param name="node">The node to evaluate.</param>
        /// <param name="correctness">The correctness score.</param>
        /// <param name="efficiency">The efficiency score.</param>
        /// <param name="robustness">The robustness score.</param>
        /// <param name="maintainability">The maintainability score.</param>
        /// <returns>The evaluated node.</returns>
        Task<ThoughtNodeAdapter> EvaluateNodeAsync(ThoughtNodeAdapter node, double correctness, double efficiency, double robustness, double maintainability);

        /// <summary>
        /// Prunes the specified node.
        /// </summary>
        /// <param name="node">The node to prune.</param>
        /// <returns>The pruned node.</returns>
        Task<ThoughtNodeAdapter> PruneNodeAsync(ThoughtNodeAdapter node);

        /// <summary>
        /// Adds metadata to the specified node.
        /// </summary>
        /// <param name="node">The node to add metadata to.</param>
        /// <param name="key">The metadata key.</param>
        /// <param name="value">The metadata value.</param>
        /// <returns>The updated node.</returns>
        Task<ThoughtNodeAdapter> AddMetadataAsync(ThoughtNodeAdapter node, string key, object value);

        /// <summary>
        /// Visualizes the tree with the specified root node.
        /// </summary>
        /// <param name="root">The root node.</param>
        /// <param name="format">The format to visualize the tree in.</param>
        /// <returns>The visualization.</returns>
        Task<string> VisualizeTreeAsync(ThoughtNodeAdapter root, string format = "text");
    }
}
